provider/webhook/webhook.go

```diff
@@ -0,0 +1,263 @@
+	}
+	if err := df.UnmarshalJSON(b); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal response body of DomainFilter: %v", err)
+	}
+
+	if contentType != mediaTypeFormatAndVersion {
+		return nil, fmt.Errorf("wrong content type returned from server: %s", contentType)
+	}
+
+	return &WebhookProvider{
+		client:          client,
+		remoteServerURL: parsedURL,
+		DomainFilter:    df,
+	}, nil
+}
+
+// Records will make a GET call to remoteServerURL/records and return the results
+func (p WebhookProvider) Records(ctx context.Context) ([]*endpoint.Endpoint, error) {
+	u := p.remoteServerURL.JoinPath("records").String()
+	req, err := http.NewRequest("GET", u, nil)
+	if err != nil {
+		recordsErrorsGauge.Inc()
+		log.Debugf("Failed to create request: %s", err.Error())
+		return nil, err
+	}
+	req.Header.Set(acceptHeader, mediaTypeFormatAndVersion)
+	resp, err := p.client.Do(req)
+	if err != nil {
+		recordsErrorsGauge.Inc()
+		log.Debugf("Failed to perform request: %s", err.Error())
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		recordsErrorsGauge.Inc()
+		log.Debugf("Failed to get records with code %d", resp.StatusCode)
+		return nil, fmt.Errorf("failed to get records with code %d", resp.StatusCode)
+	}
+
+	endpoints := []*endpoint.Endpoint{}
+	if err := json.NewDecoder(resp.Body).Decode(&endpoints); err != nil {
+		recordsErrorsGauge.Inc()
+		log.Debugf("Failed to decode response body: %s", err.Error())
+		return nil, err
+	}
+	return endpoints, nil
+}
+
+// ApplyChanges will make a POST to remoteServerURL/records with the changes
+func (p WebhookProvider) ApplyChanges(ctx context.Context, changes *plan.Changes) error {
+	u := p.remoteServerURL.JoinPath("records").String()
+
+	b := new(bytes.Buffer)
+	if err := json.NewEncoder(b).Encode(changes); err != nil {
+		applyChangesErrorsGauge.Inc()
+		log.Debugf("Failed to encode changes: %s", err.Error())
+		return err
+	}
+
+	req, err := http.NewRequest("POST", u, b)
+	if err != nil {
+		applyChangesErrorsGauge.Inc()
+		log.Debugf("Failed to create request: %s", err.Error())
+		return err
+	}
+
+	req.Header.Set(contentTypeHeader, mediaTypeFormatAndVersion)
+
+	resp, err := p.client.Do(req)
+	if err != nil {
+		applyChangesErrorsGauge.Inc()
+		log.Debugf("Failed to perform request: %s", err.Error())
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusNoContent {
+		applyChangesErrorsGauge.Inc()
+		log.Debugf("Failed to apply changes with code %d", resp.StatusCode)
+		return fmt.Errorf("failed to apply changes with code %d", resp.StatusCode)
+	}
+	return nil
+}
+
+// AdjustEndpoints will call the provider doing a POST on `/adjustendpoints` which will return a list of modified endpoints
+// based on a provider specific requirement.
+// This method returns an empty slice in case there is a technical error on the provider's side so that no endpoints will be considered.
```

Comment:
I think we might need to add an `error` to the return values of `AdjustEndpoints`. We wouldn't want a transient error to cause External-DNS to try to re-create all existing endpoints.

Reply:
I rebased to support that in the meantime, so I will add tests and code for that.